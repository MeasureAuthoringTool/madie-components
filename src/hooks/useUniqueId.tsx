declare global {
  interface Window {
    idCounter: number;
  }
}

window.idCounter = 0;

export default function useUniqueId(prefix?: string): string {
  window.idCounter = ++window.idCounter || 1;
  return `${prefix || ""}${window.idCounter}`;
}

export function resetUniqueId(): void {
  window.idCounter = 0;
}
