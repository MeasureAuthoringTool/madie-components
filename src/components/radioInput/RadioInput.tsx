import React from "react";
import tw from "twin.macro";
import useUniqueId from "../../hooks/useUniqueId";

export interface RadioInputProps
  extends React.InputHTMLAttributes<HTMLInputElement> {
  children?: never;
  id?: string;
  label: string;
  name: string;
  value: string | ReadonlyArray<string> | number;
  disabled?: boolean;
  "data-testid"?: string;
}

const StyledInput = tw.input`
  mb-1
  checked:text-primary-500
  checked:bg-primary-500
  bg-white
  disabled:bg-gray-100
  dark:bg-gray-975
  disabled:dark:bg-gray-950
  border-gray-300
  dark:border-gray-600
  disabled:dark:border-gray-700
  focus:ring-2
  focus:ring-offset-2
  focus:ring-primary-500!
  focus:dark:ring-offset-gray-975!
  `;

const StyledLabel = tw.label`
  font-body
  ml-2
  mr-5
  text-gray-800
  disabled:text-gray-500
  dark:text-gray-300`;

const DisabledLabel = tw(StyledLabel)`
  text-gray-500
  dark:text-gray-600`;

export default function RadioInput(args: RadioInputProps) {
  const uniqueId = useUniqueId("radio-input-");
  // Capture id, type and name here so that props will not contain it
  const { id, label, type, name, disabled, ...props } = args;
  const inputId = id || uniqueId;
  const labelId = `${inputId}-label`;

  const LabelEl = disabled ? DisabledLabel : StyledLabel;

  return (
    <>
      <StyledInput
        type="radio"
        id={inputId}
        name={name}
        disabled={disabled}
        {...props}
      />
      <LabelEl id={labelId} htmlFor={inputId}>
        {label}
      </LabelEl>
    </>
  );
}
