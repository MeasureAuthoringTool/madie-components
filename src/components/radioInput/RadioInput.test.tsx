import * as React from "react";
import { render } from "@testing-library/react";
import { resetUniqueId } from "../../hooks/useUniqueId";
import RadioInput from "./RadioInput";

const labelText = "Radio Label";
const formName = "radioFormName";
const formValue = "formValue";
const testId = "test-radio";
const labelId = `${testId}-label`;

describe("RadioInput component", () => {
  beforeEach(resetUniqueId);

  it("should render a RadioInput component, passing all extra properties to the input element", () => {
    const { getByTestId } = render(
      <RadioInput
        label={labelText}
        name={formName}
        value={formValue}
        data-testid={testId}
      />
    );
    const input = getByTestId(testId) as HTMLInputElement;
    expect(input).toBeInstanceOf(HTMLInputElement);
    expect(input.type).toBe("radio");
    expect(input).toBeInTheDocument();
    expect(input).toMatchSnapshot();
    expect(input.id).toBeDefined();
  });

  it("should use the ID specified", () => {
    const { getByText, getByTestId } = render(
      <RadioInput
        label={labelText}
        name={formName}
        value={formValue}
        data-testid={testId}
        id={testId}
      />
    );
    const input = getByTestId(testId) as HTMLInputElement;
    const label = getByText(labelText) as HTMLLabelElement;
    expect(input.id).toBe(testId);
    expect(label.id).toBe(labelId);
  });

  it("should be able to disable", () => {
    const { container, getByTestId } = render(
      <RadioInput
        label={labelText}
        name={formName}
        value={formValue}
        data-testid={testId}
        disabled={true}
      />
    );
    expect(container).toMatchSnapshot();
    const input = getByTestId(testId) as HTMLInputElement;
    expect(input).toBeDisabled();
  });
});
