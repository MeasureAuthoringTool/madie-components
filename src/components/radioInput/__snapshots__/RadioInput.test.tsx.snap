// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioInput component should be able to disable 1`] = `
<div>
  <input
    class="RadioInput__StyledInput-sc-jjc4eq-0 eeUXAV"
    data-testid="test-radio"
    disabled=""
    id="radio-input-1"
    name="radioFormName"
    type="radio"
    value="formValue"
  />
  <label
    class="RadioInput__StyledLabel-sc-jjc4eq-1 RadioInput__DisabledLabel-sc-jjc4eq-2 ljjYLs dpCDXv"
    for="radio-input-1"
    id="radio-input-1-label"
  >
    Radio Label
  </label>
</div>
`;

exports[`RadioInput component should render a RadioInput component, passing all extra properties to the input element 1`] = `
<input
  class="RadioInput__StyledInput-sc-jjc4eq-0 eeUXAV"
  data-testid="test-radio"
  id="radio-input-1"
  name="radioFormName"
  type="radio"
  value="formValue"
/>
`;
